### PR TITLE
Android: Added action and category as extra parameters

### PR DESCRIPTION
**Pull request by chrisism**
## Description
Extended the available parameters for starting an android activity. Added the action and category parameters to associate with the intent/app.

## Motivation and context
This changes is needed to be able to fully activate android applications from Kodi with extended parameters. Currently is only possible to simply open up a basic android app, but with this change we can activate apps with extended parameters making it possible to incorporate android apps more into Kodi.
Also since the latest Android OS updates it isn't possible to do it through normal shell/bash commands. This will solve it for many addons currently using this technique.

## How has this been tested?
System and code tests. Tested on local android emulators. Working as expected.
Needs more testing on different types of android machines and OS versions to see if it solves for all versions.

## What is the effect on users?
Won't break current usage (Matrix) of StartAndroidActivity(). But users executing Android commands can now use more parameters in their calls. All addon developers trying to execute Android commands through shell commands should now use this command through the Kodi api instead.

## Screenshots (if appropriate):
None

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by a1rwulf** [comment]
In theory extras and flags can be used with current master already, so you have to append you new parameters to the end of the list in order to not break existing clients.

**Comment by a1rwulf** [comment]
Seems you have a duplication of PR's: https://github.com/xbmc/xbmc/pull/21882

Which one is the correct one to have a look at?

**Comment by chrisism** [comment]
> In theory extras and flags can be used with current master already, so you have to append you new parameters to the end of the list in order to not break existing clients.

You are correct. My motivation was that the flags and extra parameters can consist of a bigger value/collection, where as action and category are single string values. Extras should be something like a json string for example.
So I thought it feels more correct to define the bigger string values as the later/last parameters. However, this does mean that it breaks the current definition indeed even though its still in alpha/beta phase. What shall we do? I can now easily turn it around and put them in the end.

**Comment by a1rwulf** [comment]
I get your intentions, however I'm not using a lot of addons and I'm not an active addon developer, so I can't say how fast app developers are with adding new stuff to their addons and I don't know how many addons even use this API at all.
Usually I'm rather safe then sorry and wouldn't recommend to potentially break clients just for cosmetic preferences.
@ksooo What's you opinion on that?

**Comment by chrisism** [comment]
I changed it now so that the parameters are added in the end, making it not breaking. Although from a 'cosmetic' point of view I would recommend them in previous position. But maybe indeed better safe than sorry.
@a1rwulf thanks for adding the needed labels. Hopefully with this change in I can make some of addon developers happy with how to use android from the addons.

**Review by a1rwulf — APPROVED** [review]
LGTM now.

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (a1rwulf) [inline xbmc/platform/android/activity/XBMCApp.cpp:1050]: If you add a logline here, you need to use `{}` for the if.

**Review by a1rwulf — CHANGES_REQUESTED** [review]
Make sure you format your code according to our guidelines, otherwise jenkins will complain.
The easiest way that I know to achieve this is to:
```
git add your_changed_files
git-clang-format
git add whatever_has_been_reformatted_by_last_command
git commit ...
```

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: is there a linter enabled somewhere? Been switching between my chromebook and laptop with the development for testing with android. That kicks of the builds. Fixing it now, but indeed want to see it beforehand.

On another note, I still notice some issues with the intent parameter when invoking android apps. This is separate from the actual changes in this PR, but something I noticed already with your previous PR. If the intent is used it will throw exceptions. Perhaps I'll start up a different branch to see if we can incorporate "setComponent" to solve that part.
Starting up basic android apps still works as before though.

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: You need to install git-clang-format as described in my other post.
Not sure which exception you refer to.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: The exception is not giving further details, except that it is triggered once the intent value is applied. Without it starts up the application as expected.
```
2022-09-21 15:30:15.699 T:2924    DEBUG <general>: StartActivity: package: com.retroarch
2022-09-21 15:30:15.699 T:2924    DEBUG <general>: StartActivity: intent: android.intent.action.MAIN
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: dataType: 
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: dataURI: 
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: flags: 
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: extras: [{"key": "ROM", "value": "'/storage/emulated/0/Download/work/*********.nes'", "type": "string"}, {"key": "LIBRETRO", "value": "/data/user/0/com.retroarch/cores/nestopia_libretro_android.so", "type": "string"}, {"key": "CONFIGFILE", "value": "/storage/emulated/0/Android/data/com.retroarch/files/retroarch.cfg", "type": "string"}]
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: action: com.retroarch/.browser.retroactivity.RetroActivityFuture
2022-09-21 15:30:15.701 T:2924    DEBUG <general>: StartActivity: category: android.intent.category.LAUNCHER
2022-09-21 15:30:15.712 T:2924    ERROR <general>: StartActivity: ExceptionOccurred launching com.retroarch
```

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: I think the exception should show up in logcat.
But you can add `xbmc_jnienv()->ExceptionDescribe();` between `xbmc_jnienv()->ExceptionCheck()` and `xbmc_jnienv()->ExceptionClear();` and it should show up in the log if I'm not mistaken.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Installed clang-format by the way. Did the formatting, but it had quite some formatting changes, even in code I didnt touch. So curious if it is applied correctly now

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: > Installed clang-format by the way. Did the formatting, but it had quite some formatting changes, even in code I didnt touch. 

That's why I referred to `git-clang-format` and not `clang-format` which only formats lines that you have changed.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Alright, clear. I merged in the master branch again to update everything, getting an older merge commit in. It aborts the build now. Didn't see anything about it in the contribution documentation as far as I can see.
Should I revert the last commits?

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Ok so ususally we don't merge master back in our featurebranches but instead rebase so you never get merge commits in your feature branch.
I checked out your current state and here's what I suggest:
```
git pull --rebase upstream master
git rebase -i hash-of-first-commit^ # Note the ^ which is needed to include your first commit for the interactive rebase
```
You'll get an editor, where you will delete the line with the clang-changes commit.
For the very top commit you leave the keyword "pick" for all the others you change it to "f".
Then you save and quit the editor.
Your branch will then have only one commit.
Follow up with:
```
git reset HEAD^
git add xbmc/
git-clang-format
git add xbmc/
git commit
```

This will fix your branch so that you have one commit with all your changes perfectly formatted.

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Forgot one thing:
Once you have only one commit on your branch "don't pull again", instead you need to force-push your branch so that this PR will get updated with the correct version.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Ah, overlooked your second post. Did the first and pushed that indeed.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: I should propably use git pull rebase more often. 
But yes didnt do the force push. Would it work if I made a minor change and commit,pushed that to trigger the PR update?

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Apparently something went wrong with your PR.
If you do a `git pull --rebase upstream master` your commits will be recommited on top of current master.
You must force push after that, a normal push will be rejected by git because your local branch doesn't match with your remote branch anymore.

Probably the easiest way to fix your work is to:
```
git fetch upstream
git reset --hard upstream/master
git cherry-pick bdf731d6f7c3feb3892a86e4ca69ba0e358d5708
```
You might want to change the author to yourself after picking my commit:
```
git commit --amend --author "chrisism <cwjungerius@gmail.com>"
```

Then force push your branch:
```
git push -f origin add-android-intent-parameters
```

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Mm, it cannot find that commit for the cherry pick (bad object). Did I need to redo the git pull -rebase before the git fetch?

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Ah sorry, that commit lives in my fork.
So you need to:
```
git remote add a1rwulf git@github.com/a1rwulf/xbmc.git
git fetch a1rwulf
```
Then you'll be able to pick it.

Comment (chrisism) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Did a clean git pull rebase actually and fixed all conflicts with the force push. Seems to have worked.
I still have the component/class parameter to do, but I will start a clean feature branch for that and do it the correct way.
Lets finish this one. All reviewers need to approve?

**Comment by a1rwulf** [comment]
Sorry for making you jump through all the hoops @chrisism but can you please just drop your clang format commit, we agreed to only format code that we change in PR's.
Also please change the commit message of your first commit to:
`Android: Add action and category parameter to StartActivity builtin`
Your current message `Proper commit for adding android parameters` doesn't quite nail it.
After that we can merge your PR.

**Review by garbear — APPROVED** [review]
Agree with @a1rwulf, otherwise LGTM.

**Comment by chrisism** [comment]
No problem, will do. Lets see what is the easiest way of erasing that commit?

**Comment by a1rwulf** [comment]
git reset --hard HEAD^

**Comment by chrisism** [comment]
done, thanks

**Review by a1rwulf — APPROVED** [review]
Looks good now.
Thanks for your contribution!

File: xbmc/interfaces/builtins/AndroidBuiltins.cpp
Comment (enen92) [inline xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: Sorry for the delay @chrisism but adding this changes the interface for python add-ons. Please update the description on the very same file

Comment (a1rwulf) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: Ah I missed that in my last PR too.

Comment (chrisism) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: Do we do a new PR or continue on this one?

Comment (fuzzard) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: Just PR the doc change separate

Comment (chrisism) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: And do we want a new branch for it or can I re-use the current one?
I will start a new one for a different android fix, I can also do the doc changes in that one. Let me know what is preferred.

Comment (fuzzard) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: if you've got something else you are looking to PR, just add it as a first commit in that branch, and just comment on the PR. that'll be fine.

Comment (enen92) [reply xbmc/interfaces/builtins/AndroidBuiltins.cpp:22]: Pr is merged already so you can use the same branch (after syncing with current master) or create a new one, anything will do.
Thanks and sorry for the late request

**Pull request merged**
